File: sdk/go/temporalstore/proxy_client.go
package temporalstore

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type ProxyOptions struct {
	Endpoint      string
	NamespaceName string
	TableName     string
	APIKey        string
	Timeout       time.Duration
}

type ProxyClient struct {
	endpoint string
	options  ProxyOptions
	client   *http.Client
}

func ConnectProxy(options ProxyOptions) *ProxyClient {
	timeout := options.Timeout
	if timeout == 0 {
		timeout = 5 * time.Second
	}
	return &ProxyClient{
		endpoint: strings.TrimRight(options.Endpoint, "/"),
		options:  options,
		client:   &http.Client{Timeout: timeout},
	}
}

func (c *ProxyClient) PutString(key, value string, ttlMs uint64) error {
	body := c.keyBody(key)
	body["value"] = bytesValue(value)
	if ttlMs > 0 {
		body["ttl_ms"] = ttlMs
		return c.post("/ProxyService/SetEx", body, nil)
	}
	return c.post("/ProxyService/Set", body, nil)
}

func (c *ProxyClient) GetString(key string) (string, error) {
	var data struct {
		Value string `json:"value"`
	}
	err := c.post("/ProxyService/Get", c.keyBody(key), &data)
	return data.Value, err
}

func (c *ProxyClient) DeleteObject(key string) error {
	return c.post("/ProxyService/Delete", c.keyBody(key), nil)
}

func (c *ProxyClient) Expire(key string, ttlMs uint64) error {
	body := c.keyBody(key)
	body["ttl_ms"] = ttlMs
	return c.post("/ProxyService/Expire", body, nil)
}

func (c *ProxyClient) TTL(key string) (uint64, error) {
	var data struct {
		TTL uint64 `json:"ttl_ms"`
	}
	err := c.post("/ProxyService/Ttl", c.keyBody(key), &data)
	return data.TTL, err
}

func (c *ProxyClient) HSet(key, field, value string) error {
	body := c.keyBody(key)
	body["field"] = field
	body["value"] = bytesValue(value)
	return c.post("/ProxyService/HSet", body, nil)
}

func (c *ProxyClient) HGet(key, field string) (string, error) {
	body := c.keyBody(key)
	body["field"] = field
	var data struct {
		Value string `json:"value"`
	}
	err := c.post("/ProxyService/HGet", body, &data)
	return data.Value, err
}

func (c *ProxyClient) HDel(key, field string) error {
	body := c.keyBody(key)
	body["field"] = field
	return c.post("/ProxyService/HDel", body, nil)
}

func (c *ProxyClient) SAdd(key, member string) error {
	body := c.keyBody(key)
	body["member"] = bytesValue(member)
	return c.post("/ProxyService/SAdd", body, nil)
}

func (c *ProxyClient) SMembers(key string) ([]string, error) {
	var data struct {
		Members []string `json:"members"`
	}
	err := c.post("/ProxyService/SMembers", c.keyBody(key), &data)
	return data.Members, err
}

func (c *ProxyClient) AddFeaturePoints(key string, points []FeaturePoint) error {
	body := c.keyBody(key)
	body["points"] = featurePointBodies(points)
	return c.post("/ProxyService/FeatureAdd", body, nil)
}

func (c *ProxyClient) QueryFeaturePoints(
	key string,
	startTs uint64,
	endTs uint64,
	count uint64,
	filters []FeatureFilter,
) ([]FeaturePoint, error) {
	body := c.queryBody(key, startTs, endTs, count, filters)
	var data struct {
		Points []FeaturePoint `json:"points"`
	}
	err := c.post("/ProxyService/FeatureQuery", body, &data)
	return data.Points, err
}

func (c *ProxyClient) AddSequenceFeatureRows(key string, rows []SequenceFeatureRow) error {
	body := c.keyBody(key)
	body["rows"] = sequenceRowBodies(rows)
	return c.post("/ProxyService/SequenceAdd", body, nil)
}

func (c *ProxyClient) QuerySequenceFeatureRows(
	key string,
	startTs uint64,
	endTs uint64,
	count uint64,
	filters []FeatureFilter,
) ([]SequenceFeatureRow, error) {
	body := c.queryBody(key, startTs, endTs, count, filters)
	var data struct {
		Rows []SequenceFeatureRow `json:"rows"`
	}
	err := c.post("/ProxyService/SequenceQuery", body, &data)
	return data.Rows, err
}

func (c *ProxyClient) AddIPSInstance(
	ipsTable string,
	uid int64,
	timestampUs int64,
	actionType int32,
	logicalTable int32,
	features []IpsFeatureStat,
) error {
	body := map[string]interface{}{
		"namespace":     c.options.NamespaceName,
		"table":         c.options.TableName,
		"table_name":    c.options.TableName,
		"ips_table":     ipsTable,
		"key":           ipsTable,
		"uid":           uid,
		"timestamp_us":  timestampUs,
		"timestamp_ms":  timestampUs / 1000,
		"action_type":   actionType,
		"table_id":      logicalTable,
		"logical_table": logicalTable,
		"features":      features,
	}
	instance, err := json.Marshal(features)
	if err != nil {
		return err
	}
	body["instance"] = bytesValue(string(instance))
	return c.post("/ProxyService/IpsAdd", body, nil)
}

func (c *ProxyClient) QueryIPSLastInstances(
	ipsTable string,
	uid int64,
	actionType int32,
	logicalTable int32,
	slot int32,
	topK int32,
	lastInstances int64,
) ([]IpsFeatureStat, error) {
	body := map[string]interface{}{
		"namespace":      c.options.NamespaceName,
		"table":          c.options.TableName,
		"table_name":     c.options.TableName,
		"ips_table":      ipsTable,
		"key":            ipsTable,
		"uid":            uid,
		"action_type":    actionType,
		"logical_table":  logicalTable,
		"slot":           slot,
		"top_k":          topK,
		"last_instances": lastInstances,
		"count":          lastInstances,
	}
	var data struct {
		Features []IpsFeatureStat `json:"features"`
	}
	err := c.post("/ProxyService/IpsQueryLast", body, &data)
	return data.Features, err
}

func (c *ProxyClient) RiskIncrement(
	key string,
	amount int64,
	ttlSeconds uint64,
	precision RiskPrecision,
	uuid string,
	occurTimeSeconds uint64,
) error {
	body := c.keyBody(key)
	body["amount"] = amount
	body["ttl_seconds"] = ttlSeconds
	body["ttl_ms"] = ttlSeconds * 1000
	body["precision"] = precision
	body["precision_ms"] = riskPrecisionMs(precision)
	body["uuid"] = uuid
	body["occur_time_seconds"] = occurTimeSeconds
	body["timestamp_ms"] = proxyTimestampMs(occurTimeSeconds)
	return c.post("/ProxyService/RiskIncrement", body, nil)
}

func (c *ProxyClient) RiskCount(
	key string,
	precision RiskPrecision,
	windowStart int64,
	windowEnd int64,
	windowUnit WindowUnit,
) (int64, error) {
	body := c.keyBody(key)
	body["precision"] = precision
	body["window_start"] = windowStart
	body["window_end"] = windowEnd
	body["window_unit"] = windowUnit
	startMs, endMs := riskWindowMs(windowStart, windowEnd, windowUnit)
	body["start_ms"] = startMs
	body["end_ms"] = endMs
	var data struct {
		Count int64 `json:"count"`
	}
	err := c.post("/ProxyService/RiskCount", body, &data)
	return data.Count, err
}

func (c *ProxyClient) keyBody(key string) map[string]interface{} {
	return map[string]interface{}{
		"namespace":  c.options.NamespaceName,
		"table":      c.options.TableName,
		"table_name": c.options.TableName,
		"key":        key,
	}
}

func (c *ProxyClient) queryBody(
	key string,
	startTs uint64,
	endTs uint64,
	count uint64,
	filters []FeatureFilter,
) map[string]interface{} {
	body := c.keyBody(key)
	body["start_ts"] = startTs
	body["end_ts"] = endTs
	body["start_ms"] = startTs
	body["end_ms"] = endTs
	body["count"] = count
	body["filters"] = featureFilterBodies(filters)
	return body
}

func (c *ProxyClient) post(path string, body interface{}, out interface{}) error {
	payload, err := json.Marshal(body)
	if err != nil {
		return err
	}
	req, err := http.NewRequest(http.MethodPost, c.endpoint+path, bytes.NewReader(payload))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	if c.options.APIKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.options.APIKey)
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	responseBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("temporalstore proxy http %d: %s", resp.StatusCode, string(responseBody))
	}
	if len(responseBody) == 0 {
		return nil
	}
	data, err := unwrapProxyResponse(responseBody)
	if err != nil {
		return err
	}
	if out == nil {
		return nil
	}
	if len(data) > 0 {
		return json.Unmarshal(data, out)
	}
	return nil
}

func unwrapProxyResponse(body []byte) (json.RawMessage, error) {
	var envelope struct {
		OK      *bool           `json:"ok"`
		Code    interface{}     `json:"code"`
		Message string          `json:"message"`
		Data    json.RawMessage `json:"data"`
	}
	if err := json.Unmarshal(body, &envelope); err == nil && envelope.OK != nil {
		if !*envelope.OK {
			return nil, fmt.Errorf("temporalstore proxy code %v: %s", envelope.Code, envelope.Message)
		}
		return envelope.Data, nil
	}

	var executed struct {
		Status struct {
			OK      bool   `json:"ok"`
			Code    string `json:"code"`
			Message string `json:"message"`
		} `json:"status"`
		Response json.RawMessage `json:"response"`
	}
	if err := json.Unmarshal(body, &executed); err == nil && len(executed.Response) > 0 {
		if !executed.Status.OK {
			return nil, fmt.Errorf("temporalstore proxy code %s: %s", executed.Status.Code, executed.Status.Message)
		}
		return flattenCommandResponse(executed.Response)
	}

	return body, nil
}

func flattenCommandResponse(raw json.RawMessage) (json.RawMessage, error) {
	var response map[string]json.RawMessage
	if err := json.Unmarshal(raw, &response); err != nil {
		return nil, err
	}
	var kind string
	_ = json.Unmarshal(response["kind"], &kind)
	switch kind {
	case "empty":
		return []byte("{}"), nil
	case "bytes":
		var value []int
		if len(response["value"]) > 0 && !bytes.Equal(response["value"], []byte("null")) {
			_ = json.Unmarshal(response["value"], &value)
		}
		payload := map[string]interface{}{"value": intsToString(value)}
		return json.Marshal(payload)
	case "integer", "aggregate":
		var value int64
		_ = json.Unmarshal(response["value"], &value)
		payload := map[string]interface{}{"value": value, "count": value, "ttl_ms": uint64(value)}
		return json.Marshal(payload)
	case "members":
		var members [][]int
		_ = json.Unmarshal(response["members"], &members)
		values := make([]string, 0, len(members))
		for _, member := range members {
			values = append(values, intsToString(member))
		}
		return json.Marshal(map[string]interface{}{"members": values})
	case "feature_points":
		return flattenFeaturePoints(response["points"])
	case "sequence_rows":
		return flattenSequenceRows(response["rows"])
	default:
		return raw, nil
	}
}

func featurePointBodies(points []FeaturePoint) []map[string]interface{} {
	out := make([]map[string]interface{}, 0, len(points))
	for _, point := range points {
		out = append(out, map[string]interface{}{
			"timestamp":    point.Timestamp,
			"timestamp_ms": point.Timestamp,
			"value":        bytesValue(point.Value),
		})
	}
	return out
}

func sequenceRowBodies(rows []SequenceFeatureRow) []map[string]interface{} {
	out := make([]map[string]interface{}, 0, len(rows))
	for _, row := range rows {
		out = append(out, map[string]interface{}{
			"timestamp":    row.Timestamp,
			"timestamp_ms": row.Timestamp,
			"gid":          row.GID,
			"action_type":  row.ActionType,
			"duration":     row.Duration,
			"author_id":    row.AuthorID,
		})
	}
	return out
}

func featureFilterBodies(filters []FeatureFilter) []map[string]interface{} {
	out := make([]map[string]interface{}, 0, len(filters))
	for _, filter := range filters {
		out = append(out, map[string]interface{}{
			"field": filter.Field,
			"op":    featureFilterOpName(filter.Op),
			"value": filter.Value,
		})
	}
	return out
}

func featureFilterOpName(op FeatureFilterOp) string {
	switch op {
	case FeatureFilterNotEqual:
		return "not_equal"
	case FeatureFilterGreaterThan:
		return "greater_than"
	case FeatureFilterLessThan:
		return "less_than"
	default:
		return "equal"
	}
}

func flattenFeaturePoints(raw json.RawMessage) (json.RawMessage, error) {
	var points []map[string]json.RawMessage
	_ = json.Unmarshal(raw, &points)
	outPoints := make([]map[string]interface{}, 0, len(points))
	features := make([]IpsFeatureStat, 0)
	for _, point := range points {
		var timestamp uint64
		_ = json.Unmarshal(point["timestamp_ms"], &timestamp)
		var value []int
		_ = json.Unmarshal(point["value"], &value)
		text := intsToString(value)
		outPoints = append(outPoints, map[string]interface{}{"timestamp": timestamp, "timestamp_ms": timestamp, "value": text})
		var decoded []IpsFeatureStat
		if err := json.Unmarshal([]byte(text), &decoded); err == nil {
			features = append(features, decoded...)
		}
	}
	return json.Marshal(map[string]interface{}{"points": outPoints, "features": features})
}

func flattenSequenceRows(raw json.RawMessage) (json.RawMessage, error) {
	var rows []map[string]interface{}
	_ = json.Unmarshal(raw, &rows)
	for _, row := range rows {
		if timestamp, ok := row["timestamp_ms"]; ok {
			row["timestamp"] = timestamp
		}
	}
	return json.Marshal(map[string]interface{}{"rows": rows})
}

func proxyTimestampMs(occurTimeSeconds uint64) uint64 {
	if occurTimeSeconds == 0 {
		return uint64(time.Now().UnixMilli())
	}
	return occurTimeSeconds * 1000
}

func riskPrecisionMs(precision RiskPrecision) uint64 {
	switch precision {
	case RiskOneSecond:
		return 1000
	case RiskFiveSeconds:
		return 5000
	case RiskTenSeconds:
		return 10000
	case RiskOneMinute:
		return 60000
	case RiskFiveMinutes:
		return 5 * 60000
	case RiskTenMinutes:
		return 10 * 60000
	case RiskOneHour:
		return 60 * 60000
	case RiskOneDay:
		return 24 * 60 * 60000
	case RiskOneMonth:
		return 30 * 24 * 60 * 60000
	default:
		return 60000
	}
}

func riskWindowMs(windowStart int64, windowEnd int64, unit WindowUnit) (uint64, uint64) {
	now := time.Now().UnixMilli()
	end := windowEnd
	if end <= 0 {
		end = now
	}
	start := windowStart
	if start < 0 {
		start = end - int64(windowUnitMs(unit))
	}
	if start < 0 {
		start = 0
	}
	return uint64(start), uint64(end)
}

func windowUnitMs(unit WindowUnit) uint64 {
	switch unit {
	case WindowSecond:
		return 1000
	case WindowMinute:
		return 60000
	case WindowDay:
		return 24 * 60 * 60000
	default:
		return 60 * 60000
	}
}

func bytesValue(value string) []int {
	out := make([]int, 0, len(value))
	for _, b := range []byte(value) {
		out = append(out, int(b))
	}
	return out
}

func intsToString(value []int) string {
	out := make([]byte, 0, len(value))
	for _, b := range value {
		out = append(out, byte(b))
	}
	return string(out)
}

File: sdk/python/temporalstore/proxy_client.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 MatrixArkAI
from __future__ import annotations

import json
import urllib.error
import urllib.request
from dataclasses import asdict, dataclass
from typing import Any, Dict, Iterable, List, Optional

from .client import (
    FeatureFilter,
    FeatureFilterOp,
    FeaturePoint,
    RiskPrecision,
    SequenceFeatureRow,
    TemporalStoreError,
    WindowUnit,
)


@dataclass
class ProxyOptions:
    endpoint: str
    namespace_name: str
    table_name: str
    api_key: str = ""
    timeout_seconds: float = 5.0


class ProxyClient:
    def __init__(self, options: ProxyOptions):
        self.options = options
        self.endpoint = options.endpoint.rstrip("/")

    def put_string(self, key: str, value: str, ttl_ms: Optional[int] = None) -> None:
        body: Dict[str, Any] = self._key_body(key)
        body["value"] = _bytes_value(value)
        if ttl_ms is not None:
            body["ttl_ms"] = ttl_ms
            self._post("/ProxyService/SetEx", body)
            return
        self._post("/ProxyService/Set", body)

    def get_string(self, key: str) -> str:
        data = self._post("/ProxyService/Get", self._key_body(key))
        return str(data.get("value", ""))

    def delete_object(self, key: str) -> None:
        self._post("/ProxyService/Delete", self._key_body(key))

    def expire(self, key: str, ttl_ms: int) -> None:
        body = self._key_body(key)
        body["ttl_ms"] = ttl_ms
        self._post("/ProxyService/Expire", body)

    def ttl(self, key: str) -> int:
        data = self._post("/ProxyService/Ttl", self._key_body(key))
        return int(data.get("ttl_ms", 0))

    def hset(self, key: str, field: str, value: str) -> None:
        body = self._key_body(key)
        body.update({"field": field, "value": _bytes_value(value)})
        self._post("/ProxyService/HSet", body)

    def batch_hset(self, entries: Iterable[Dict[str, Any]]) -> None:
        grouped: Dict[str, List[List[Any]]] = {}
        for entry in entries:
            key = str(entry["key"])
            grouped.setdefault(key, []).append(
                [str(entry["field"]), _bytes_value(str(entry.get("value", "")))]
            )
        for key, kv in grouped.items():
            body = self._key_body(key)
            body["entries"] = kv
            self._post("/ProxyService/HMSet", body)

    def hget(self, key: str, field: str) -> str:
        body = self._key_body(key)
        body["field"] = field
        data = self._post("/ProxyService/HGet", body)
        return str(data.get("value", ""))

    def hdel(self, key: str, field: str) -> None:
        body = self._key_body(key)
        body["field"] = field
        self._post("/ProxyService/HDel", body)

    def sadd(self, key: str, member: str) -> None:
        body = self._key_body(key)
        body["member"] = _bytes_value(member)
        self._post("/ProxyService/SAdd", body)

    def smembers(self, key: str) -> List[str]:
        data = self._post("/ProxyService/SMembers", self._key_body(key))
        return [str(item) for item in data.get("members", [])]

    def add_feature_points(self, key: str, points: Iterable[FeaturePoint]) -> None:
        body = self._key_body(key)
        body["points"] = [_feature_point_body(point) for point in points]
        self._post("/ProxyService/FeatureAdd", body)

    def query_feature_points(
        self,
        key: str,
        start_ts: int,
        end_ts: int,
        count: int,
        filters: Iterable[FeatureFilter] = (),
    ) -> List[FeaturePoint]:
        body = self._query_body(key, start_ts, end_ts, count, filters)
        data = self._post("/ProxyService/FeatureQuery", body)
        return [
            FeaturePoint(int(row.get("timestamp", row.get("timestamp_ms", 0))), _string_value(row.get("value", "")))
            for row in data.get("points", [])
        ]

    def add_sequence_feature_rows(self, key: str, rows: Iterable[SequenceFeatureRow]) -> None:
        body = self._key_body(key)
        body["command"] = {"kind": "sequence_add", "key": key, "rows": [_sequence_row_body(row) for row in rows]}
        self._post("/ProxyService/ExecuteTableCmd", body)

    def query_sequence_feature_rows(
        self,
        key: str,
        start_ts: int,
        end_ts: int,
        count: int,
        filters: Iterable[FeatureFilter] = (),
    ) -> List[SequenceFeatureRow]:
        body = self._key_body(key)
        body["command"] = {
            "kind": "sequence_query",
            "key": key,
            "start_ms": start_ts,
            "end_ms": end_ts,
            "count": int(count),
            "filters": [_feature_filter_body(item) for item in filters],
        }
        data = self._post("/ProxyService/ExecuteTableCmd", body)
        return [
            SequenceFeatureRow(
                int(row.get("timestamp", row.get("timestamp_ms", 0))),
                int(row["gid"]),
                int(row["action_type"]),
                int(row["duration"]),
                int(row["author_id"]),
            )
            for row in data.get("rows", [])
        ]



    def risk_increment(
        self,
        key: str,
        amount: int = 1,
        ttl_seconds: int = 24 * 3600,
        precision: RiskPrecision = RiskPrecision.ONE_MINUTE,
        uuid: str = "",
        occur_time_seconds: int = 0,
    ) -> None:
        body = self._key_body(key)
        body.update(
            {
                "amount": amount,
                "ttl_seconds": ttl_seconds,
                "precision": int(precision),
                "uuid": uuid,
                "occur_time_seconds": occur_time_seconds,
            }
        )
        body["command"] = {
            "kind": "control_state_increment",
            "key": key,
            "timestamp_ms": int(occur_time_seconds * 1000) if occur_time_seconds else 0,
            "amount": int(amount),
        }
        self._post("/ProxyService/ExecuteTableCmd", body)

    def risk_count(
        self,
        key: str,
        precision: RiskPrecision = RiskPrecision.ONE_MINUTE,
        window_start: int = -1,
        window_end: int = 0,
        window_unit: WindowUnit = WindowUnit.HOUR,
    ) -> int:
        body = self._key_body(key)
        body.update(
            {
                "precision": int(precision),
                "window_start": window_start,
                "window_end": window_end,
                "window_unit": int(window_unit),
            }
        )
        body["command"] = {"kind": "control_state_count", "key": key, "start_ms": int(window_start), "end_ms": int(window_end)}
        data = self._post("/ProxyService/ExecuteTableCmd", body)
        return int(data.get("count", 0))

    def feature_agg_query(
        self,
        key: str,
        start_ts: int,
        end_ts: int,
        aggregator: str = "count",
        count: Optional[int] = None,
    ) -> int:
        """Exact serving-time aggregate (count/sum/min/max/avg/first/last)."""
        body = self._key_body(key)
        body.update({"start_ms": int(start_ts), "end_ms": int(end_ts), "aggregator": str(aggregator)})
        if count is not None:
            body["count"] = int(count)
        data = self._post("/ProxyService/FeatureAggQuery", body)
        return int(data.get("value", data.get("count", 0)))

    def control_state_increment(
        self,
        key: str,
        amount: int = 1,
        ttl_seconds: int = 24 * 3600,
        precision: RiskPrecision = RiskPrecision.ONE_MINUTE,
        uuid: str = "",
        occur_time_seconds: int = 0,
    ) -> None:
        """Canonical name for the renamed Control State counter increment."""
        self.risk_increment(key, amount, ttl_seconds, precision, uuid, occur_time_seconds)

    def control_state_count(
        self,
        key: str,
        precision: RiskPrecision = RiskPrecision.ONE_MINUTE,
        window_start: int = -1,
        window_end: int = 0,
        window_unit: WindowUnit = WindowUnit.HOUR,
    ) -> int:
        """Canonical name for the renamed Control State windowed count."""
        return self.risk_count(key, precision, window_start, window_end, window_unit)

    def _key_body(self, key: str) -> Dict[str, Any]:
        return {
            "namespace": self.options.namespace_name,
            "table": self.options.table_name,
            "table_name": self.options.table_name,
            "key": key,
        }

    def _query_body(
        self,
        key: str,
        start_ts: int,
        end_ts: int,
        count: int,
        filters: Iterable[FeatureFilter],
    ) -> Dict[str, Any]:
        body = self._key_body(key)
        body.update(
            {
                "start_ts": start_ts,
                "end_ts": end_ts,
                "start_ms": start_ts,
                "end_ms": end_ts,
                "count": count,
                "filters": [_feature_filter_body(item) for item in filters],
            }
        )
        return body

    def _post(self, path: str, body: Dict[str, Any]) -> Dict[str, Any]:
        request = urllib.request.Request(
            self.endpoint + path,
            data=json.dumps(body).encode("utf-8"),
            headers=self._headers(),
            method="POST",
        )
        try:
            with urllib.request.urlopen(request, timeout=self.options.timeout_seconds) as response:
                payload = json.loads(response.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            message = exc.read().decode("utf-8", errors="replace")
            raise TemporalStoreError(exc.code, message) from exc
        except urllib.error.URLError as exc:
            raise TemporalStoreError(0, str(exc)) from exc

        return _unwrap_proxy_response(payload)

    def _headers(self) -> Dict[str, str]:
        headers = {"Content-Type": "application/json"}
        if self.options.api_key:
            headers["Authorization"] = "Bearer " + self.options.api_key
        return headers


def _bytes_value(value: str) -> List[int]:
    return list(value.encode("utf-8"))


def _string_value(value: Any) -> str:
    if isinstance(value, list):
        return bytes(int(item) & 0xFF for item in value).decode("utf-8", errors="replace")
    return str(value)


def _feature_filter_body(value: FeatureFilter) -> Dict[str, Any]:
    names = {
        FeatureFilterOp.EQUAL: "equal",
        FeatureFilterOp.NOT_EQUAL: "not_equal",
        FeatureFilterOp.GREATER_THAN: "greater_than",
        FeatureFilterOp.LESS_THAN: "less_than",
    }
    return {"field": value.field, "op": names.get(value.op, "equal"), "value": int(value.value)}


def _feature_point_body(value: FeaturePoint) -> Dict[str, Any]:
    return {"timestamp_ms": int(value.timestamp), "value": _bytes_value(value.value)}


def _sequence_row_body(value: SequenceFeatureRow) -> Dict[str, Any]:
    return {
        "timestamp_ms": int(value.timestamp),
        "gid": int(value.gid),
        "action_type": int(value.action_type),
        "duration": int(value.duration),
        "author_id": int(value.author_id),
    }


def _unwrap_proxy_response(payload: Any) -> Dict[str, Any]:
    if not isinstance(payload, dict):
        return {}
    if "ok" in payload:
        if not payload.get("ok", False):
            code = payload.get("code", 0)
            raise TemporalStoreError(int(code) if isinstance(code, int) else 0, str(payload.get("message", "")))
        data = payload.get("data", {})
        return data if isinstance(data, dict) else {}
    if "status" in payload and "response" in payload:
        status = payload.get("status") or {}
        if not status.get("ok", False):
            raise TemporalStoreError(0, str(status.get("message", status.get("code", ""))))
        return _flatten_command_response(payload.get("response", {}))
    return payload


def _flatten_command_response(response: Any) -> Dict[str, Any]:
    if not isinstance(response, dict):
        return {}
    kind = response.get("kind")
    if kind == "empty":
        return {}
    if kind == "bytes":
        return {"value": _string_value(response.get("value", []))}
    if kind in {"integer", "aggregate"}:
        value = int(response.get("value", 0))
        return {"value": value, "count": value, "ttl_ms": value}
    if kind == "members":
        return {"members": [_string_value(item) for item in response.get("members", [])]}
    if kind == "feature_points":
        return {"points": [_feature_point_result(item) for item in response.get("points", [])]}
    if kind == "sequence_rows":
        return {"rows": [_sequence_row_result(item) for item in response.get("rows", [])]}
    return response


def _feature_point_result(value: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "timestamp": int(value.get("timestamp", value.get("timestamp_ms", 0))),
        "value": _string_value(value.get("value", "")),
    }


def _sequence_row_result(value: Dict[str, Any]) -> Dict[str, Any]:
    out = dict(value)
    out["timestamp"] = int(out.get("timestamp", out.get("timestamp_ms", 0)))
    return out
